python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,520 @@
+
+    @property
+    def width(self):
+        return self._width
+
+    def apply(self, value):
+        if value is None:
+            return None
+        if self._type == IntegerType or self._type == LongType:
+            return value - (((value % self._width) + self._width) % self._width)
+        elif self._type == StringType or self._type == BinaryType:
+            return value[0 : min(self._width, len(value))]
+        else:  # decimal case
+            unscaled_value = Transform._unscale_decimal(value)
+            applied_value = unscaled_value - (
+                ((unscaled_value % self._width) + self._width) % self._width
+            )
+            return Decimal(f"{applied_value}e{value.as_tuple().exponent}")
+
+    def can_transform(self, target: Type) -> bool:
+        return self._type == target
+
+    def preserve_order(self) -> bool:
+        return True
+
+    def satisfy_order(self, other: Transform) -> bool:
+        if self == other:
+            return True
+        elif (
+            StringType == self._type
+            and isinstance(other, Truncate)
+            and StringType == other._type
+        ):
+            return self._width >= other._width
+
+        return False
+
+    def to_human_string(self, value) -> str:
+        if value is None:
+            return "null"
+
+        return (
+            (base64.b64encode(value)).decode("ISO-8859-1")
+            if self._type == BinaryType
+            else str(value)
+        )
+
+
+class UnknownTransform(Transform):
+    def __init__(self, transform_type: Type, transform: str):
+        super().__init__(
+            transform,
+            f"UnknownTransform(transform_type={repr(transform_type)}, transform='{transform}')",
+        )
+        self._type = transform_type
+        self._transform = transform
+
+    def apply(self, value):
+        raise AttributeError(f"Cannot apply unsupported transform: {self.__str__()}")
+
+    def can_transform(self, target: Type) -> bool:
+        return repr(self._type) == repr(target)
+
+    def get_result_type(self, source: Type) -> Type:
+        return StringType
+
+
+VoidTransform = Transform("void", "transforms.always_null()")
+
+_HAS_WIDTH = re.compile("(\\w+)\\[(\\d+)\\]")
+_TIME_TRANSFORMS = {
+    DateType: {
+        "year": Time(DateType, "year"),
+        "month": Time(DateType, "month"),
+        "day": Time(DateType, "day"),
+    },
+    TimestampType: {
+        "year": Time(TimestampType, "year"),
+        "month": Time(TimestampType, "month"),
+        "day": Time(TimestampType, "day"),
+        "hour": Time(TimestampType, "hour"),
+    },
+    TimestamptzType: {
+        "year": Time(TimestamptzType, "year"),
+        "month": Time(TimestamptzType, "month"),
+        "day": Time(TimestamptzType, "day"),
+        "hour": Time(TimestamptzType, "hour"),
+    },
+}
+
+
+def from_string(transform_type: Type, transform: str) -> Transform:
+    match = _HAS_WIDTH.match(transform)
+
+    if match is not None:
```

Comment:
Rather than calling lower in several places, I think this should call `transform.lower()` just once and use the result.